File: frontend/src/app/models/media.ts
import {FileFolderInfo} from './filefolderinfo';

export interface Download {
  id: number;
  path: string;
  file_name: string;
  file_hash: string;
  size: number;
  resolution: number; // e.g., 1080, 2160
  file_format: string; // e.g., "mp4", "mkv"
  video_format: string; // e.g., "h264", "h265"
  audio_format: string; // e.g., "aac", "ac3"
  audio_language: string | null; // e.g., "eng", "tel"
  subtitle_format: string | null; // e.g., "srt", "ass"
  subtitle_language: string | null; // e.g., "eng"
  duration: number; // in seconds
  youtube_id: string;
  youtube_channel: string;
  file_exists: boolean;
  profile_id: number; // ID of the TrailerProfile used
  media_id: number;
  added_at: Date; // When trailer was downloaded
  updated_at: Date; // When file was last modified
}

export function parseDate(dateStr: string): Date | null {
  if (!dateStr) {
    return null; // Return epoch if dateStr is empty or invalid
  }
  if (dateStr.endsWith('Z')) {
    return new Date(dateStr);
  }
  return new Date(`${dateStr}Z`);
}

export function mapDownload(download: any): Download {
  return {
    ...download,
    file_exists: Boolean(download.file_exists),
    added_at: parseDate(download.added_at),
    updated_at: parseDate(download.updated_at),
  };
}

/** A media item with a trailer download currently in flight (runtime
 * state from the backend's in-memory registry — never stored in the DB). */
export interface InflightDownload {
  media_id: number;
  profile_id: number;
}

/** List-level status is computed from downloads + monitor (the backend
 * stores no status since v0.11.0). 'downloading' comes only from the
 * runtime in-flight overlay (MediaService.downloadingResource). */
export function computeMediaStatus(monitor: boolean, downloads: Download[], downloading = false): string {
  if (downloading) return 'downloading';
  if (downloads.some((download) => download.file_exists)) return 'downloaded';
  if (monitor) return 'monitored';
  return 'missing';
}

export function buildDownloadMap(downloads: Download[]): Map<number, Download[]> {
  const downloadMap = new Map<number, Download[]>();

  // Group downloads by media_id
  downloads.forEach((download) => {
    if (!downloadMap.has(download.media_id)) {
      downloadMap.set(download.media_id, []);
    }
    downloadMap.get(download.media_id)!.push(download);
  });

  return downloadMap;
}

export interface Media {
  is_movie: boolean;
  connection_id: number;
  arr_id: number;
  title: string;
  clean_title: string;
  year: number;
  language: string;
  studio: string;
  media_exists: boolean;
  media_filename: string;
  season_count: number;
  overview: string;
  runtime: number;
  youtube_trailer_id: string;
  folder_path: string;
  imdb_id: string;
  txdb_id: string;
  tmdb_id: number | null;
  tvdb_id: number | null;
  title_slug: string;
  poster_url: string;
  fanart_url: string;
  poster_path: string;
  fanart_path: string;
  monitor: boolean;
  arr_monitored: boolean;
  status: string;
  id: number;
  added_at: Date;
  updated_at: Date;
  downloaded_at: Date;
  downloads: Download[];
  files: FileFolderInfo | null;

  plex_rating_key: string | null;
  plex_connection_id: number | null;
  plex_trailer: boolean | null;

  // Additional properties
  isImageLoaded: boolean;
}

export function mapMedia(media: any): Media {
  return {
    ...media,
    is_movie: Boolean(media.is_movie),
    media_exists: Boolean(media.media_exists),
    monitor: Boolean(media.monitor),
    arr_monitored: Boolean(media.arr_monitored),
    // Phase 5: the API no longer sends a status field — this is only a
    // first-paint value until the downloads map arrives and
    // computeMediaStatus takes over.
    status: media.monitor ? 'monitored' : 'missing',
    added_at: parseDate(media.added_at),
    updated_at: parseDate(media.updated_at),
    downloaded_at: parseDate(media.downloaded_at),
    isImageLoaded: false,
  };
}

export interface SearchMedia {
  id: number;
  title: string;
  year: number;
  youtube_trailer_id: string;
  imdb_id: string;
  txdb_id: string;
  tmdb_id: number | null;
  tvdb_id: number | null;
  is_movie: boolean;
  poster_path: string | null;
}

export interface FolderInfo {
  type: string;
  name: string;
  size: string;
  path: string;
  files: FolderInfo[];
  modified: Date;
  isExpanded: boolean;
}

export function mapFolderInfo(folder: any): FolderInfo {
  let _files = [
    {
      type: 'folder',
      name: 'None',
      size: '',
      path: '',
      files: [],
      modified: new Date(),
      isExpanded: false,
    },
  ];
  if (folder.files) {
    _files = folder.files.map((file: any) => (isFile(file) ? mapFileInfo(file) : mapFolderInfo(file)));
  }
  return {
    ...folder,
    isExpanded: false,
    modified: new Date(`${folder.created}`),
    files: _files,
  };
}

function mapFileInfo(file: any): FolderInfo {
  return {
    ...file,
    isExpanded: false,
    modified: new Date(`${file.created}`),
  };
}

function isFile(file: any): boolean {
  // Implement this function based on how you differentiate between a file and a folder in your data.
  // For example, you might check a 'type' property:
  return file.type === 'file';
}

File: backend/core/base/database/models/media.py
from datetime import datetime, timezone
from enum import Enum
from sqlalchemy import Boolean, Column, String, text, Enum as sa_Enum
from sqlmodel import Field, Integer, SQLModel


def get_current_time():
    return datetime.now(timezone.utc)


def get_current_year():
    return datetime.now(timezone.utc).year


class MonitorStatus(Enum):
    """Monitor status for media. \n"""

    DOWNLOADED = "downloaded"
    DOWNLOADING = "downloading"
    MISSING = "missing"
    MONITORED = "monitored"


class MediaBase(SQLModel):
    """Base class for the Media model. \n
    Note: \n
        🚨**DO NOT USE THIS CLASS DIRECTLY.**🚨 \n
    Use MediaCreate, MediaRead, MediaUpdate models instead.
    """

    connection_id: int
    arr_id: int = Field(index=True)
    is_movie: bool = Field(default=True, index=True)
    title: str = Field(index=True)
    clean_title: str = Field(
        default="",
        sa_column=Column(
            String, server_default=text("('')"), index=True, nullable=False
        ),
    )
    year: int = Field(default_factory=get_current_year, index=True)
    language: str = Field(default="en", index=True)
    studio: str = Field(
        default="",
        sa_column=Column(String, server_default=text("('')"), nullable=False),
    )
    media_exists: bool = Field(
        default=False,
        sa_column=Column(Boolean, server_default="0", nullable=False),
    )
    media_filename: str = Field(
        default="",
        sa_column=Column(String, server_default=text("('')"), nullable=False),
    )
    season_count: int = Field(
        default=0,
        sa_column=Column(Integer, server_default="0", nullable=False),
    )
    overview: str | None = None
    runtime: int = 0
    # website: str | None = None
    youtube_trailer_id: str | None = None
    folder_path: str | None = None
    imdb_id: str | None = Field(default=None, index=True)
    txdb_id: str = Field(index=True)
    title_slug: str = Field(
        default="",
        sa_column=Column(
            String, server_default=text("('')"), index=True, nullable=False
        ),
    )
    poster_url: str | None = None
    fanart_url: str | None = None
    poster_path: str | None = None
    fanart_path: str | None = None
    trailer_exists: bool = Field(default=False)
    monitor: bool = Field(default=False)
    arr_monitored: bool = Field(default=False)
    status: MonitorStatus = Field(
        default=MonitorStatus.MISSING,
        sa_column=Column(
            sa_Enum(MonitorStatus, native_enum=False),
            server_default=text("'MISSING'"),
            nullable=False,
        ),
    )


class Media(MediaBase, table=True):
    """Media model for the database. \n
    This class is used for database CRUD operations only \n
    Note: \n
        🚨**DO NOT USE THIS CLASS DIRECTLY.**🚨 \n
    Use MediaCreate, MediaRead, MediaUpdate models instead.
    """

    id: int | None = Field(default=None, primary_key=True)
    connection_id: int = Field(
        foreign_key="connection.id", index=True, ondelete="CASCADE"
    )
    is_movie: bool = Field(default=True, index=True)

    added_at: datetime = Field(default_factory=get_current_time)
    updated_at: datetime = Field(default_factory=get_current_time)
    downloaded_at: datetime | None = Field(default=None)


class MediaCreate(MediaBase):
    """Media model for creating a new media objects. \n
    Defaults:
    - is_movie: True
    - year: current year
    - language: "en"
    - runtime: 0
    - youtube_trailer_id: None
    - trailer_exists: False
    - monitor: False
    - arr_monitored: False
    """

    pass


class MediaRead(MediaBase):
    """Media model for reading media."""

    id: int
    added_at: datetime
    updated_at: datetime
    downloaded_at: datetime | None


class MediaUpdate(MediaBase):
    """Media model for updating media. \n
    Defaults:
    - updated_at: current time [if any field is updated]
    """

    connection_id: int | None = None  # type: ignore
    arr_id: int | None = None  # type: ignore
    title: str | None = None  # type: ignore
    year: int | None = None  # type: ignore
    language: str | None = None  # type: ignore
    runtime: int | None = None  # type: ignore
    txdb_id: str | None = None  # type: ignore
    trailer_exists: bool | None = None  # type: ignore
    monitor: bool | None = None  # type: ignore
    arr_monitored: bool | None = None  # type: ignore

    downloaded_at: datetime | None = None
    updated_at: datetime = Field(default_factory=get_current_time)
